cleaner.go

```diff
@@ -106,31 +227,53 @@ func (c Cleaner) printStats() {
 }
 
 func (c *Cleaner) RemoveRelease(releaseName string) error {
-	log.Printf("Looking for %s release...", releaseName)
-	release, err := c.helmClient.GetRelease(releaseName)
-	if err == driver.ErrReleaseNotFound {
-		log.Printf("%s release not found, nothing to do", releaseName)
-		return nil
-	}
-	if err != nil {
-		return err
-	}
+	done := make(chan bool)
+	var errs []error
+	go func() {
+		hasResourcesToCheck := true
+		for hasResourcesToCheck {
+			hasResourcesToCheck = false
+			errs = []error{}
+			for _, r := range resources[releaseName] {
+				ro := &unstructured.Unstructured{}
+				ro.SetGroupVersionKind(r.GetObjectKind().GroupVersionKind())
+				if err := c.k8sCli.Get(context.Background(), types.NamespacedName{Name: r.GetName(), Namespace: r.GetNamespace()}, ro); kerrors.IsNotFound(err) || meta.IsNoMatchError(err) {
+					continue
+				} else if err != nil {
+					errs = append(errs, gerr.Wrap(err, "getting resource"))
+				} else {
+					if len(ro.GetFinalizers()) != 0 {
+						rodc := ro.DeepCopy()
+						rodc.SetGroupVersionKind(r.GetObjectKind().GroupVersionKind())
+						rodc.SetFinalizers([]string{})
+						if err := c.k8sCli.Update(context.Background(), rodc); err != nil {
+							errs = append(errs, gerr.Wrap(err, "failed patching"))
+						}
+					}
+				}
+				if err := c.k8sCli.Delete(context.Background(), r); kerrors.IsNotFound(err) || meta.IsNoMatchError(err) {
+					continue
+				} else if err != nil {
+					errs = append(errs, gerr.Wrap(err, fmt.Sprintf("%T %v", err, "failed deleting")))
+				} else {
+					log.Printf("deleting resource %v %v/%v\n", r.GetObjectKind().GroupVersionKind(), r.GetNamespace(), r.GetName())
+					hasResourcesToCheck = true
+				}
+			}
+		}
+		done <- true
+	}()
 
-	log.Printf("Found %s release in the namespace %s: status %s", release.Name, release.Namespace, release.Info.Status.String())
-	log.Println(" Uninstalling...")
-	err = c.helmClient.UninstallRelease(&helmclient.ChartSpec{
-		ReleaseName:  releaseName,
-		DisableHooks: true,
-		Wait:         true,
-		Timeout:      time.Minute,
-		Force:        true,
-	})
-	if err != nil {
-		return err
+	select {
+	case <-done:
+		if len(errs) != 0 {
+			return errors.NewAggregate(errs)
+		}
+		return nil
+	case <-time.After(10 * time.Minute):
+		errs = append(errs, fmt.Errorf("deleting %v timed out after 30 minutes", releaseName))
+		return errors.NewAggregate(errs)
```

Comment:
it usually takes less than a minute to purge a release, 10 minutes should be plenty. In case it times out, it will forward the latest set of errors.